File: front-end/lib/domain/entities/eleve_model.dart
class Eleve {
  final String id;
  final String email;
  final String nom;
  final String prenom;
  final int age;
  final DateTime createdAt;
  final DateTime updatedAt;

  Eleve({
    required this.id,
    required this.email,
    required this.nom,
    required this.prenom,
    required this.age,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Eleve.fromJson(Map<String, dynamic> json) {
    return Eleve(
      id: json['_id'],
      email: json['email'],
      nom: json['nom'],
      prenom: json['prenom'],
      age: json['age'],
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'email': email,
      'nom': nom,
      'prenom': prenom,
      'age': age,
    };
  }
}

File: back-end/models/eleve.model.js
const mongoose = require('mongoose');

const eleveSchema = new mongoose.Schema({
  email: {
    type: String,
    required: true,
    unique: true,
    trim: true,
  },
  password: {
    type: String,
    required: true,
  },
  nom: {
    type: String,
    required: true,
    trim: true,
  },
  prenom: {
    type: String,
    required: true,
    trim: true,
  },
  age: {
    type: Number,
    required: true,
    min: 5,
    max: 18,
  },
  niveau: {
    type: Number,
    default: 0,
  },
  niveauxCompletes: {
    type: [Number],
    default: []
  },
  accessibleLevels: {
    type: [Boolean],
    default: function () {
      // By default, only level 0 is accessible
      return [true, false, false, false, false, false, false, false, false, false, false, false, false, false];
    }
  },
  completedLevels: {
    type: [Boolean],
    default: function () {
      // By default, no levels are completed
      return [false, false, false, false, false, false, false, false, false, false, false, false, false, false];
    }
  },
  createdAt: {
    type: Date,
    default: Date.now
  },
  updatedAt: {
    type: Date,
    default: Date.now
  }
}, {
  timestamps: true,
});

module.exports = mongoose.model('Eleve', eleveSchema);
